KIPs/kip-81.md

```diff
@@ -0,0 +1,773 @@
+    mapping(address => address) private nodeIdToVoter; // nodeId => voterAddr
+    mapping(address => address) private voterToNodeId; // voterAddr => nodeId
+
+    /// @dev Creates a new Tracker and populate initial values from AddressBook
+    function createTracker(uint256 trackStart, uint256 trackEnd)
+        external returns(uint256 trackerId);
+
+    /// @dev Re-evaluate Tracker contents related to the staking contract
+    /// Anyone can call this function, but `staking` must be a staking contract
+    /// registered to the AddressBook.
+    function refreshStake(address staking) external;
+
+    /// @dev Re-evaluate voter account mapping related to the staking contract
+    /// Anyone can call this function, but `staking` must be a staking contract
+    /// registered to the AddressBook.
+    function refreshVoter(address staking) external;
+
+    /// @dev Return integer fields of a tracker
+    function getTrackerSummary(uint256 trackerId) external view returns(
+        uint256 trackStart,
+        uint256 trackEnd,
+        uint256 numNodes,
+        uint256 totalVotes,
+        uint256 eligibleNodes);
+
+    /// @dev Return balances and votes of all nodes stored in a tracker
+    function getAllTrackedNodes(uint256 trackerId) external view returns(
+        address[] memory nodeIds,
+        uint256[] memory nodeBalances,
+        uint256[] memory nodeVotes);
+
+    /// @dev Return the balance and votes of a specified node
+    function getTrackedNode(uint256 trackerId, address nodeId) external view returns(
+        uint256 nodeBalance,
+        uint256 nodeVotes);
+
+    /// @dev Find the node who owns the given voter account
+    function getNodeFromVoter(address voter) external view returns(address nodeId);
+
+    /// @dev Find the voter account of the given node
+    function getVoterFromNode(address nodeId) external view returns(address voter);
+}
+```
+
+#### Usage
+
+A voting contract can utilize StakingTracker as follows.
+
+1. When a governance proposal is submitted, the voting contract calls `createTracker()` to finalize eligible GC nodes list and evaluate voting powers.
+2. Before `trackEnd` block, GCs stake or unstake their KLAYs from their CnStakingV2 contracts. The CnStakingV2 contracts will then call `refreshStake()` to notify balance change.
+3. GCs may change their voter account in their CnStakingV2 contracts. The CnStakingV2 contracts will then call `refreshVoter()` to notify voter account change.
+4. After `trackEnd` block, the voting powers are frozen. The voting contract use StakingTracker getters to process votes casted by voter accounts.
+
+#### Example implementation
+
+Below is an example implementation of `createTracker()`, `refreshStake()`, and `refreshVoter()` functions.
+
+```solidity
+abstract contract StakingTracker {
+    function createTracker(uint256 trackStart, uint256 trackEnd)
+        external override returns(uint256 trackerId)
+    {
+        trackerId = allTrackerIds.length + 1;
```

Comment:
trackerId starts from 1, not zero?

Reply:
This particular example does start from 1 to align with proposalId.

However, the current KIP spec doesn't specify which number `trackerId` must start from. I think it's up to the implementation.